File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(numbers, target, k, total):
    global answer
    
    if k == len(numbers):
        if total == target:
            answer += 1
        return 
    
    dfs(numbers, target, k + 1, total + numbers[k])
    dfs(numbers, target, k + 1, total - numbers[k])
    
    
def solution(numbers, target):
    global answer
    
    answer = 0
    dfs(numbers, target, 0, 0)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    int n, target, answer;
    int[] numbers;
    
    public int solution(int[] numbers, int target) {
        n = numbers.length;
        this.target = target; this.numbers = numbers;
        
        answer = 0;
        dfs(0, 0);
        return answer;
    }
    
    public void dfs(int k, int total) {
        if (k == n) {
            if (total == target) {
                answer++;
            }
            return;
        }
        dfs(k + 1, total + numbers[k]);
        dfs(k + 1, total - numbers[k]);
    }
}